### PR TITLE
Add Django migration operations for changing Elasticsearch indexes

### PR DESCRIPTION
**Pull request by joxl**
## Technical Summary

Adds four new Django migration utilities for managing Elasticsearch index changes:

- `CreateIndex` -- migration operation for creating an Elasticsearch index.
- `DeleteIndex` -- migration operation for deleting an Elasticsearch index.
- `UpdateIndexMapping` -- migration operation for updating an existing Elasticsearch index mapping.
- `make_elastic_migration` -- manage command for creating Django migration modules using the above three operations.

The `UpdateIndexMapping` operation has the ability to apply a partial mapping. This is new functionality, and is required by #32393.

Additional implementation details and explanations can be found in the updates to [README.rst](https://github.com/dimagi/commcare-hq/blob/jm/elastic-index-migrations/corehq/apps/es/README.rst).

Review: commit by commit

Jira: [SAAS-14003](https://dimagi-dev.atlassian.net/browse/SAAS-14003)

### Safety story

New code is fully tested, some new tests added for existing code. This change set is mostly isolated to Elastic management operations, does not impact data, and should not affect any user-facing application logic.

### Automated test coverage

All new code is fully tested. Some existing code has more tests. Also includes removal of untested code.

### QA Plan

QA ticket: [QAV2-4350](https://dimagi-dev.atlassian.net/browse/QAV2-4350)

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

[SAAS-14003]: https://dimagi-dev.atlassian.net/browse/SAAS-14003?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[QAV2-4350]: https://dimagi-dev.atlassian.net/browse/QAV2-4350?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: corehq/apps/es/README.rst [hunk 1]
```diff
@@ -1,6 +1,174 @@
...
+  mapping in Elasticsearch has diverged from the mapping in code.
+
+Because CommCare HQ Elastic mappings have been able to drift between
+environments, it is no longer possible to update some index mappings using the
+current technique. On some indexes, the live index mappings have sufficiently
+diverged that there is no common, "full mapping definition" that can be applied
+on all environments. This means that in order to push mapping changes to all
+environments, new mapping update logic is needed which is capable of updating
+individual properties on an Elastic index mapping while leaving other (existing)
+properties unchanged.
```
Comment (millerdev) [inline corehq/apps/es/README.rst:87]: Seems like this paragraph will become outdated as soon as we adopt the new migrations framework. What do you think of rephrasing it to be a perspective on past practice rather than "current" state?

I'm thinking specifically of this phrase:

> it is no longer possible to update some index mappings using the current technique

Might be as simple as using the word "historical" rather than "current"?

File: corehq/apps/es/README.rst [hunk 2]
```diff
@@ -1,6 +1,174 @@
...
+2. Export the mappings from a deployed environment into a local file.
+3. Compare the two files.
+
+In practice, this might look like the following example:
+
+.. code-block:: shell
+
+   ./manage.py print_elastic_mappings sms --no-names > ./sms-in-code.py
+   cchq <env> django-manage print_elastic_mappings smslogs_2020-01-28:sms --no-names > ./sms-live.py
+   diff -u ./sms-live.py ./sms-in-code.py
```
Comment (millerdev) [inline corehq/apps/es/README.rst:124]: :star:

File: corehq/apps/es/README.rst [hunk 3]
```diff
@@ -1,6 +1,174 @@
...
+items (index and tunning settings values) are configurable, as well as what
+default tuning settings will be used when not customized by the environment.
+
+**Important note**: These Elasticsearch index tuning settings are not "live".
+That is: changing their values on a deployed environment will not have any
+immediate affect on live indexes in Elasticsearch. Instead, these values are
+only ever used when an index is created (for example, during a fresh CommCare HQ
+installation or when an existing index is reindexed into a new one). This means
+that making new values become "live" requires a update to the CommCare HQ
+codebase.
```
Comment (millerdev) [inline corehq/apps/es/README.rst:166]: Can this be a bit more specific about what kind of update would be required to do that? I think it would involve an index migration, but having that stated explicitly here might be valuable.

File: corehq/apps/es/README.rst [hunk 4]
```diff
@@ -1,6 +1,174 @@
...
+
+In order to guarantee stability (or "manageability", if you will) of this core
+data, it is important that Elasticsearch indexes are maintained in a consistent
+state across all environments as a concrete design feature of CommCare HQ. This
+design constraint is accomplished by managing Elasticsearch index modifications
+(for example: creating indexes, updating index mappings, etc) exclusively
+through Django's migration framework. This ensures that all Elasticsearch index
+modifications will be part of standard CommCare HQ code deployment procedures,
+thereby preventing Elasticsearch index state drift between maintained CommCare
+HQ deployments.
```
Comment (millerdev) [inline corehq/apps/es/README.rst:22]: Would it be worth adding a paragraph here to the effect of

> One or more migrations are required any time index mappings, tuning parameters, ... are changed in code.

That list should be expanded to include all things in code that describe the state of elastic entities where the code change would not immediately take effect on deployment without a corresponding migration being applied.

File: corehq/apps/es/index/analysis.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+
```
Comment (millerdev) [inline corehq/apps/es/index/analysis.py:1]: Would it be worth adding a docstring here to note that changes to analyzers must be accompanied by a migration to apply the change to elastic?

File: corehq/apps/es/index/analysis.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+DEFAULT_ANALYZER = {
```
Comment (millerdev) [inline corehq/apps/es/index/analysis.py:2]: What do you think of renaming this to `LOWERCASE_WHITESPACE_ANALYZER`?

Maybe in a separate PR if it seems out of scope for this one?

File: corehq/apps/es/index/settings.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+IndexSettingsKey = _IndexSettingsKey()
+IndexTuningKey = _IndexTuningKey()
+
+
+# Allow removing settings from defaults by setting the value to this constant in
+# settings.ES_SETTINGS.
+# TODO: change this constant to an object that isn't actually a meaningful
+# (e.g. `__do_not_set__`) instead of `None` (because `None` is how you revert a
+# setting to the value the server is configured to use).
+REMOVE_SETTING = None
```
Comment (millerdev) [inline corehq/apps/es/index/settings.py:128]: Why is changing this left as a TODO?

Additionally, `None` is a common and easily referenced value. What do you think of using something a little harder to get by accident? (This doesn't need to be JSON-serializable, right? If it does, then `"__do_not_set__"` seems reasonable.)
```suggestion
REMOVE_SETTING = object()
```

File: corehq/apps/es/tests/utils.py [hunk 8]
```diff
@@ -51,7 +57,8 @@ class ElasticTestMixin(object):
     def setUpClass(cls):
         super().setUpClass()
         cls._es_instance = get_es_new()
-        initialize_index_and_mapping(cls._es_instance, TEST_INDEX_INFO)
+        with ignore_index_settings_key_warning:
```
Comment (millerdev) [inline corehq/apps/es/tests/utils.py:60]: This will eventually be removed, right? Maybe add a comment here to note that?

**Review by millerdev — COMMENTED** [review]
Review through eda32f0c588568fbee2e4c5db19dba366da87e10. I plan to continue next week.

File: corehq/apps/es/index/settings.py (hunk 7 above)
Comment (joxl) [reply corehq/apps/es/index/settings.py:128]: It _does_ need to be JSON-serializable (e.g. [elasticsearch.yml](https://github.com/dimagi/commcare-cloud/blob/master/environments/staging/elasticsearch.yml))

File: corehq/apps/es/README.rst (hunk 1 above)
Comment (joxl) [reply corehq/apps/es/README.rst]: See fe7b237e490

File: corehq/apps/es/README.rst (hunk 3 above)
Comment (joxl) [reply corehq/apps/es/README.rst]: See fe7b237e490

File: corehq/apps/es/README.rst (hunk 4 above)
Comment (joxl) [reply corehq/apps/es/README.rst:22]: See fe7b237e490

File: corehq/apps/es/index/analysis.py (hunk 5 above)
Comment (joxl) [reply corehq/apps/es/index/analysis.py:2]: See b40e15fa23f

File: corehq/apps/es/index/analysis.py (hunk 6 above)
Comment (joxl) [reply corehq/apps/es/index/analysis.py]: See 491c3f066a8

File: corehq/apps/es/tests/utils.py (hunk 8 above)
Comment (joxl) [reply corehq/apps/es/tests/utils.py:61]: See c585b24d403

File: corehq/apps/es/migration_operations.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        super().__init__(self.run, self.reverse_run)
+        self.name = name
+        self.type = type_
+        self.mapping = mapping
+        self.analysis = analysis
+        self.settings_key = settings_key
+        self.comment = comment
+
+    def run(self, *args, **kw):
```
[the remark this answers is not in the document]
Comment (joxl) [reply corehq/apps/es/migration_operations.py:66]: This becomes the superclass's (`RunPython`) `code()` method, which gets called with migration context args during migrations. These methods never use the args, though.

File: corehq/apps/es/tests/test_migration_operations.py [hunk 10]
```diff
@@ -311,6 +296,213 @@ def test_describe(self):
...
+        self.assertEqual(
+            manager.index_get_mapping(self.index, self.type)["properties"]["prop"],
+            {"type": "string"},
+        )
+        migration = TestMigration(UpdateIndexMapping(
+            self.index,
+            self.type,
+            {"prop": {"type": "integer"}},
+        ))
+        with self.assertRaises(RequestError) as mock:
```
[the remark this answers is not in the document]
Comment (joxl) [reply corehq/apps/es/tests/test_migration_operations.py]: See b5ef2ab82a0

File: corehq/apps/es/tests/test_migration_operations.py [hunk 11]
```diff
@@ -311,6 +296,213 @@ def test_describe(self):
...
+    """
+    class mock_datetime:
+
+        isoformat_called_count = 0
+        utcnow_called_count = 0
+
+        @classmethod
+        def isoformat(cls, value):
+            cls.isoformat_called_count += 1
+            return value
```
[the remark this answers is not in the document]
Comment (joxl) [reply corehq/apps/es/tests/test_migration_operations.py:643]: No, not to these tests. Tests that depend on the actual value mock it differently (see [here](https://github.com/dimagi/commcare-hq/blob/b5ef2ab82a0/corehq/apps/es/tests/test_command_make_elastic_migration.py#L20-L26))

File: corehq/apps/es/migration_operations.py [hunk 12]
```diff
@@ -170,24 +176,63 @@ def __init__(self, name, type_, properties, comment=None):
             ``mapping._meta.comment`` property. If ``None`` (the default), then
             the existing ``mapping._meta.comment`` value (if one exists) is
             retained.
+        :param print_diff: Optional (default ``True``) set to ``False`` to
```
[the remark this answers is not in the document]
Comment (joxl) [reply corehq/apps/es/migration_operations.py:211]: It does make it more verbose. It also replicates the default behavior of the manage command it replaces (`update_es_mapping`). That being said, the "existing default behavior" that this replicates is relatively new, and was only used in production once (in other words, it's not a long-standing precedent). This is a bit tricky because exactly "who wants to see this diff" probably depends on who's running `./manage.py migration`, rather than "who is creating the migration". I would be happy to offline about these details.

File: corehq/ex-submodules/pillowtop/migrations/0006_add_geopoint_to_case_search_index.py [hunk 13]
```diff
@@ -1,13 +1,15 @@
...
+    has been removed because it was merged further in the past than our
+    downstream hosters support window, making it safe to skip from now on.
+    """
 
     dependencies = [
         ('pillowtop', '0005_kafkacheckpoint_doc_modification_time'),
     ]
 
-    operations = [update_es_mapping(CASE_SEARCH_ALIAS)]
+    operations = []
```
[the remark this answers is not in the document]
Comment (joxl) [reply corehq/ex-submodules/pillowtop/migrations/0006_add_geopoint_to_case_search_index.py:15]: Do you know of an example for how this was doing in the past? I looked into this and found several migrations with empty operations lists. It certainly appears as though there's a precedent to remove the operations completely.

File: corehq/pillows/management/commands/print_elastic_mappings.py [hunk 14]
```diff
@@ -45,23 +45,28 @@ def add_arguments(self, parser):
...
-        parser.add_argument("cname", metavar="INDEX", choices=sorted(iter_index_cnames()),
-            help="print mapping for %(metavar)s")
-
-    def handle(self, cname, **options):
+        parser.add_argument("mapping_ident", metavar="ID", help=(
+            "Print the mapping for %(metavar)s. To print the mapping from "
+            "code, specify an index canonical name (e.g. 'cases', 'sms', etc). "
+            "To print the mapping fetched from Elasticsearch, specify a "
+            "colon-delimited index name and index '_type' (e.g. "
+            "'hqcases_2016-03-04:case', 'smslogs_2020-01-28:sms', etc)."
```
[the remark this answers is not in the document]
Comment (joxl) [reply corehq/pillows/management/commands/print_elastic_mappings.py:57]: The existing index details is the only mode that accepts the `name:type_` syntax. The "from code" mode only takes a canonical name (not `type_`).

File: corehq/apps/es/tests/test_command_make_elastic_migration.py [hunk 15]
```diff
@@ -138,10 +135,9 @@ def test_build_migration_fails_for_multiple_operations_on_the_same_index(self):
         deletes = [conflict_index]
         command = Command()
         command.empty = False
-        with self.assertRaises(CommandError) as mock:
+        prefix = f"Multiple operations for the same index ({conflict_index}):"
+        with self.assertRaisesRegex(CommandError, f"^{re.escape(prefix)}.*"):
```
Comment (millerdev) [inline corehq/apps/es/tests/test_command_make_elastic_migration.py:139]: nit: `.*` is unnecessary
```suggestion
        with self.assertRaisesRegex(CommandError, f"^{re.escape(prefix)}"):
```
Also applies to other tests using the same pattern.

File: corehq/ex-submodules/pillowtop/migrations/0006_add_geopoint_to_case_search_index.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/ex-submodules/pillowtop/migrations/0006_add_geopoint_to_case_search_index.py:15]: Is this enough to get you started?
https://github.com/dimagi/commcare-hq/blob/9fcbacf7a0f73346ae6c9238be0ccd9741c86940/corehq/util/django_migrations.py#L131-L134

File: corehq/pillows/management/commands/print_elastic_mappings.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/pillows/management/commands/print_elastic_mappings.py:57]: Oh, I thought `type_` was the same as *cannonical name*.

[the remark this answers is not in the document]
Comment (joxl) [reply corehq/pillows/management/commands/print_elastic_mappings.py:57]: You only need a canonical name to reference a document adapter. The `name:type_` syntax allows this manage command to fetch a mapping without caring which (if any) document adapter uses that index.

File: corehq/apps/es/migration_operations.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/es/migration_operations.py:211]: Fine to leave as is if you think it's valuable output for the person running `migrate`.

File: corehq/ex-submodules/pillowtop/migrations/0006_add_geopoint_to_case_search_index.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (joxl) [reply corehq/ex-submodules/pillowtop/migrations/0006_add_geopoint_to_case_search_index.py]: See 910aad67775

File: corehq/apps/es/tests/test_command_make_elastic_migration.py (hunk 15 above)
Comment (joxl) [reply corehq/apps/es/tests/test_command_make_elastic_migration.py]: See 910aad67775

**Review by millerdev — APPROVED** [review]

File: corehq/apps/es/client.py [hunk 16]
```diff
@@ -981,10 +984,8 @@ def from_python(self, doc):
     # meta methods and Elastic index read methods (pass-through on the primary
     # adapter)
     @property
-    def settings(self):
```
Comment (AmitPhulera) [inline corehq/apps/es/client.py:984]: Don't we need this property for Multiplexed Adapter?

Comment (joxl) [reply corehq/apps/es/client.py:984]: No, the settings no longer live as a "static" property on the adapter. Instead, the adapter keeps a "settings_key", which is used to render the settings data when it is needed.

**Review by AmitPhulera — APPROVED** [review]
⭐

**Review by snopoke — COMMENTED** [review]
I didn't do a thorough review but these changes look great. Thanks for the hard work on it and for all the docstrings + actual docs!

Also, TESTS!!!

**Comment by calellowitz** [comment]
Thanks for doing this! I do have one question about how this works for third party hosters, specifically around the tuning parameters you mention. Tuning parameters like number of shards per index or number of replicas will likely differ for each hoster, and require action at different times than it does for us.

Am I reading correctly that these hosters will need to create migrations, and therefore fork the codebase, to do any of this maintenance, and that they will therefore need to maintain that fork, and deal with migration conflicts each time an index changes or we do maintenance, since their migration table will be out of sync with the main codebase?

**Comment by joxl** [comment]
> and therefore fork the codebase, to do any of this maintenance                
                                                                               
Third party hosters can accomplish this by recreating the index from primary data, as you explained [here](https://dimagi.slack.com/archives/C4Q5D8ZDZ/p1641836055013600).

**Comment by snopoke** [comment]
> > and therefore fork the codebase, to do any of this maintenance
> 
> Third party hosters can accomplish this by recreating the index from primary data, as you explained [here](https://dimagi.slack.com/archives/C4Q5D8ZDZ/p1641836055013600).

Thanks for raising this @calellowitz, I would agree that it's an issue if 3rd party hosters can't tune ES without making code changes. Obviously it's possible they could use ES endpoints directly but that puts them at a significant disadvantage. We could also develop external tooling e.g. in commcare-cloud but that just creates a parallel system which Dimagi would never use but must maintain. 

To be clear I think this only applies to these 'tuning' settings which I believe are only required when creating a new index (with the exception of `number_of_replicas` which as you correctly point out should be removed from HQ and left in commcare-cloud).

In terms of reindexing from primary data - is that process going to change as well or are we keeping the existing process? It may work for most cases as @calellowitz pointed out on Slack but there is at least 1 (possibly 2) 3rd party systems that we know of which are too large for that approach.

It also still 'feels' strange to me that we are managing environment specific settings with Django migrations.

**Comment by snopoke** [comment]
> Obviously it's possible they could use ES endpoints directly but that puts them at a significant disadvantage. We could also develop external tooling e.g. in commcare-cloud but that just creates a parallel system which Dimagi would never use but must maintain.

One other point I think is relevant here is that if a 3rd party were to use the ES APIs directly or via some other tooling we provide, there would still be no good way for them to do the reindex while the system is still in use (while data is still be updated / created). The multiplex adapter pattern also requires code changes so a 3rd part wouldn't be able to use it easily.
